core/dbt/parser/schemas.py

```diff
@@ -1042,8 +1046,42 @@ def parse_metric(self, unparsed: UnparsedMetric) -> ParsedMetric:
             self.schema_parser.manifest,
             package_name,
         )
-        model_ref = "{{ " + unparsed.model + " }}"
-        get_rendered(model_ref, ctx, parsed, capture_macros=True)
+        if unparsed.type == 'ratio':
```

Comment:
Heard. If we want the validation to happen in Python code (and it should), I think we'll need to teach some of it to `dbt-core` (or a `dbt-core` dependency). I don't mind `ratio` being a special type for now.

Is there a more general solution that looks like "super"/"sub" metric types? Like super/subclasses for Python objects. Or maybe we don't want to build the semantic layer as OOP :)

Reply:
It feels wrong to me to have this code be in the parser. I'd be more comfortable if we could move some of it into the  UnparsedMetric or ParsedMetric or a subclass based on type... Having to render it does complicate things though. Is it actually necessary to render the numerator and denominator at parse time? Can they actually contain macros?  Because if they can't contain macros, I would think the get_rendered is a no-op.

Reply:
@gshank I think we do need to render the numerator and denominator at parse-time because we need to capture the `metric()` call in order to build the edges between the two metrics. In this way, `metric()` is kind of like `ref()`, but for metrics :)

I'd be ok with moving this logic into the ParsedMetric (or similar)... is it ok if we call `get_rendered` in there? I don't want to explode the number of places that we render jinja code, but i definitely hear what you're saying